File: src/scripts/lua/spells/item/74222_perform_speech.lua
--[[
  ArcEmu MMORPG Server
  Copyright (C) 2005-2007 Ascent Team <http://www.ascentemu.com/>
  Copyright (C) 2008-2022 Arcemu Team <http://www.ArcEmu.org/>
 
  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU Affero General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  any later version.
 
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Affero General Public License for more details.
 
  You should have received a copy of the GNU Affero General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
]]

-- Spawns High Tinker Mekkatorque who makes a different speech for each target NPC
function PerformSpeech_DummyHandler( effect, spell )
	local target = spell:GetTarget()
	local player = spell:GetCaster()
	
	-- Target should be an NPC we used the item on
	if target ~= NIL
	then
		local npc = target				
		local name = npc:GetName()

		if name == "Milli Featherwhistle" then
			local tinker = npc:SpawnCreature( 39712, -5451.6, -619.33, 394.55, 1.7269, 875, 10000 )
			tinker:EventChat(12, 0, "They make take our lives, but they'll never take our...", 2000)
			tinker:EventChat(12, 0, "...INNOVATION!", 4000)
			npc:EventChat(12, 0, "What? I don't even know what you're talking about! That's terrible!", 6000)
			
			player:MarkQuestObjectiveAsComplete( 25283, 1 )
			
		elseif name == "Ozzie Togglevolt"
		then
			local tinker = npc:SpawnCreature( 39712, -5496.3984, -457.6682, 395.4378, 3.2318, 875, 10000 )
			tinker:EventChat(12, 0, "What I want out of each and every one of you is a hard-target search of every refueling station, residence, warehouse, farmhouse, henhouse, outhouse, and doghouse in this area", 2000)
			tinker:EventChat(12, 0, "Your fugitive's name is Mekgineer Thermaplugg.", 4000)
			tinker:EventChat(12, 0, "Go get him!", 6000)
			npc:EventChat(12, 0, "Hmm, I suppose it could work. But it could really use a little more umph!", 8000)
			
			player:MarkQuestObjectiveAsComplete( 25283, 0 )
			
		elseif name == "Tog Rustsprocket"
		then
			local tinker = npc:SpawnCreature( 39712, -5643.3579, -525.8497, 403.9703, 4.9502, 875, 10000 )
			tinker:EventChat(12, 0, "We will not go quietly into the night! We will not vanish without a fight!", 2000)
			tinker:EventChat(12, 0, "We're going to live on! We're going to survive! Today we celebrate...", 4000)
			tinker:EventChat(12, 0, "Our Autonomy Day!", 6000)
			npc:EventChat(12, 0, "Horrible! Well, all right, maybe it just needs a little cleaning up?", 8000)
			
			player:MarkQuestObjectiveAsComplete( 25283, 2 )
		else
			spell:Cancel()
		end
		
	end
end

RegisterDummySpell( 74222, "PerformSpeech_DummyHandler" );

File: src/scripts/pythonscripts/prod/74222_perform_speech.py
#
# ArcEmu MMORPG Server
# Copyright (C) 2008-2023 <http://www.ArcEmu.org/>
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
# 
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
# 
# TODO: Missing visuals
#

import math
import arcemu

MEKKATORQUE_SPEECH_NPC_ID = 39712

PLAYBACK_SPEECH_SPELL_ID = 74209
PERFORM_SPEECH_SPELL_ID = 74222

NPC_OZZIE_TOGGLEVOLT = 1268
NPC_TOG_RUSTSPROCKET = 6119
NPC_MILLI_FEATHERWHISTLE = 7955

def PerformSpeech_handleDummyEffect( effectIndex, spell ):
	playerCaster = spell.getPlayerCaster()
	unitTarget = spell.getUnitTarget()

	if playerCaster is None:
		return True
		
	if unitTarget is None:
		return True
		
	creature = unitTarget.toCreature()
	if creature is None:
		return True
		
	creatureId = creature.getId()

	# Check if we cast the spell on the right NPC
	if creatureId != NPC_OZZIE_TOGGLEVOLT and creatureId != NPC_TOG_RUSTSPROCKET and creatureId != NPC_MILLI_FEATHERWHISTLE:
		spell.cancel()
		return True

	# Spawn Mekkatorque
	mapMgr = creature.getMapMgr()
	x = creature.getPositionX()
	y = creature.getPositionY()
	o = creature.getOrientation()
	spawnArc = ( 2 * math.pi ) - ( ( 2 * math.pi ) / 10.0 )
	
	x = x + 2 * math.cos( o + spawnArc )
	y = y + 2 * math.sin( o + spawnArc )
	
	spawn = mapMgr.spawnCreature( MEKKATORQUE_SPEECH_NPC_ID, x, y, creature.getPositionZ() )
	if spawn is not None:
		spawn.setScale( 0.5 )
		spawn.setUInt64Value( arcemu.UNIT_FIELD_CREATEDBY, playerCaster.getGUID() )
		spawn.setUInt64Value( arcemu.UNIT_FIELD_SUMMONEDBY, playerCaster.getGUID() )
		spawn.setUInt64Value( arcemu.UNIT_CREATED_BY_SPELL, PLAYBACK_SPEECH_SPELL_ID )
		spawn.faceUnit( unitTarget )
		
		# Sets the NPC who listens to the speech as target, so that the AI script will know who the speech is for
		spawn.setTargetGUID( unitTarget.getGUID() )
		
	return True

arcemu.RegisterDummySpellHandler( PERFORM_SPEECH_SPELL_ID, PerformSpeech_handleDummyEffect )
